Cuda quantized tensors, support for quantize per tensor

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59701
* #58245
* **#59700**

Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

HDCharles committed:
Cuda quantized tensors, support for quantize per tensor

Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -328,6 +329,82 @@ def test_qtensor_dtypes(self):
+            scales = torch.rand(2).abs()
+            zero_points = (torch.rand(2) * 10).round().to(zero_type)
+
+            # dqr = torch.quantize_per_channel(r, scales, zero_points, axis, dtype)
+            # dqr = dqr.to(device)
+            # dqr_cuda = torch.quantize_per_channel(r.to(device), scales.to(
+            #     device), zero_points.to(device), axis, dtype)
+            # dqr_cuda = dqr_cuda.to('cpu')
+            # self.assertEqual('cuda', dqr.device.type)
+            # self.assertEqual('cpu', dqr_cuda.device.type)
```
why comment these out?

HDCharles replied:
I split the pr into multiple parts, so the tests would fail at this point since this pr doesn't have the per channel functionality implemented

raghuramank100 replied:
Second jerry's comment, will be good to have a per-channel test and per-tensor test for .to device functionality.

HDCharles replied:
ok, I can do this.

supriyar commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -328,6 +329,82 @@ def test_qtensor_dtypes(self):
             rqr = qr.dequantize()
             self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / scale))
 
+    @unittest.skipIf(not TEST_CUDA, "No gpu is not available.")
```
nit: No gpu is available

supriyar commented on aten/src/ATen/native/quantized/TensorFactories.cpp:
```diff
@@ -94,19 +94,39 @@ Tensor empty_per_channel_affine_quantized_other_backends_stub(
 // Create an empty quantized Tensor with size, based on the options
 // and quantization parameters of the input quantized Tensor
-Tensor empty_quantized(IntArrayRef size, const Tensor& qtensor) {
+Tensor empty_quantized(
```
Previously this function used to get the tensor options from the qtensor. Now it seems like it can be overridden. What is the main motivation for having these extra options?

Also, can we update the comment above this function?

HDCharles replied:
it still has its previous functionality, but in order to copy to a new device, we have to be able to overwrite the device option. I added the other options for completeness. Without that it wouldn't be possible to make .to('cuda') or .to('cpu') work for quantized tensors

supriyar commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -34,12 +34,17 @@ static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, b
+       if (self.is_quantized()){
+         auto r = at::empty_quantized(self.sizes(), self, options);
+         r.copy_(self, non_blocking);
+         return r;
```
can we move `return r` outside the if-else?

HDCharles replied:
it throws an error about r not being defined. I originally wanted to move both return and copy outside the if statement.

jerryzh168 replied:
you can first declare the Tensor before the branch I think:
```python
Tensor r;
if (...) {
    r = ...
} else {
   r = ...
}
...
```

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

jerryzh168 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -328,6 +329,82 @@ def test_qtensor_dtypes(self):
 
+    @unittest.skipIf(not TEST_CUDA, "No gpu is available.")
+    def test_qtensor_to_device(self):
+        dtype_and_zero_types = [
```
nit: you can use itertools.product here I think: https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L559

HDCharles replied:
I need to exclude specific combinations which makes a cross product problematic

jerryzh168 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -328,6 +329,82 @@ def test_qtensor_dtypes(self):
+            scales = torch.rand(2).abs()
+            zero_points = (torch.rand(2) * 10).round().to(zero_type)
+
+            # dqr = torch.quantize_per_channel(r, scales, zero_points, axis, dtype)
```
nit: maybe we can split the quantize_per_tensor and quantize_per_channel tests to two to make it clearer?

HDCharles replied:
this is just testing to .to device functionality, its not actually testing the different quantization methods.

jerryzh168 reviewed (approved):
Looks good, had a few nit comments inline

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

raghuramank100 commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -34,10 +34,15 @@ static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, b
-                                 options.memory_format(c10::nullopt).pinned_memory(pin_out));
+      Tensor r;
+      if (self.is_quantized()) {
+        r = at::empty_quantized(self.sizes(), self, options);
```
Why is pinning not supported here?

raghuramank100 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -328,6 +329,82 @@ def test_qtensor_dtypes(self):
+                    dqtr = qtr.dequantize()
+                    qtr_cuda = torch.quantize_per_tensor(r.to(device), scale, zero_point, dtype)
+                    dqtr_cuda = qtr_cuda.dequantize()
+                    self.assertEqual(qtr, qtr_cuda)
```
Since you are trying with random data on different cores, I wonder if assertEqual will always work. Do you need a check for closeness here?, i.e something like
```
     self.assertTrue(np.allclose(qr.int_repr(), ref_res))
     self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / np.min(scales.numpy())))
```

HDCharles replied:
I can make that change but as it stands the implementations are exactly equal on cuda and cpu.

raghuramank100 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -328,6 +329,82 @@ def test_qtensor_dtypes(self):
+                # self.assertEqual(qr.int_repr(), qr_cuda.int_repr())
+                # self.assertEqual(dqr, dqr_cuda)
+
+                if zero_type == torch.long:  # per_tensor doesn't have support for Float zero_point
```
Can you exclude float zero_type in the dtype_and_zero_types and remove this check?

HDCharles replied:
if I'm going to split per_channel and per_tensor, this won't be a problem hereafter.

raghuramank100 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -615,7 +692,7 @@ def test_qtensor_copy(self):
             device = 'cpu'
             # check copy from non-quantized to quantized
-            r = torch.randn([numel], dtype=torch.float).to(device)
+            r = torch.randn([numel], dtype=torch.float, device=device)
```
Do you still need device to be 'cpu' here?

HDCharles replied:
yeah Its the same as the test in line 214 i.e. test_qtensor_float_assignment, the functionality for assigning a float value to an element of a quantized tensor and having it automatically inserted in a quantized form. 

Both this test and the one on 214 rely on that functionality which isn't currently supported on cuda.

raghuramank100 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -720,47 +797,52 @@ def test_qtensor_view(self):
+            # torch.equal is not supported for the cuda backend
+            if device == 'cpu':
+                self.assertFalse(torch.equal(b, c))
+            else:
```
what are you checking here for cuda?. Is it just checking that torch.equal fails?. Since you have already checked the fields (scale, zero point and int repr), you could remove lines from 832 to 835

HDCharles replied:
test_qtensor_view had already been updated for tests on cuda and used this method so I duplicated it for consistency. I can remove it in both cases though.

raghuramank100 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -861,19 +944,6 @@ def test_choose_qparams(self, X, reduce_range):
         np.testing.assert_array_almost_equal(X_scale, qparams[0], decimal=3)
         self.assertEqual(X_zp, qparams[1])
 
-    @unittest.skipIf(not torch.cuda.is_available() or TEST_WITH_ROCM, 'CUDA is not available')
```
This test seems to be similar to the numerics check that you have earlier, and only checks for approximate equality.

HDCharles replied:
it tests that the int representations are equal but the decimal representations are not required to be. I can certainly make the change though.

raghuramank100 commented on aten/src/ATen/native/quantized/TensorFactories.cpp:
```diff
@@ -94,23 +94,45 @@ Tensor empty_per_channel_affine_quantized_other_backends_stub(
+    c10::optional<ScalarType> dtype,
+    c10::optional<Layout> layout,
+    c10::optional<Device> device,
+    c10::optional<bool> pin_memory,
```
It looks like this function allows for quantized tensors to be pinned: Can we extend the test here to test for this? Currently it checks to see 

``` python
def test_cuda_quantization_does_not_pin_memory(self):
        # Context - https://github.com/pytorch/pytorch/issues/41115
        x = torch.randn(3)
        self.assertEqual(x.is_pinned(), False)
        q_int = torch.randint(0, 100, [1, 2, 3], device="cuda", dtype=torch.uint8)
        q = torch._make_per_tensor_quantized_tensor(q_int, scale=0.1, zero_point=0)
        x = torch.randn(3)
        self.assertEqual(x.is_pinned(), False)
```

HDCharles replied:
there's a test for pinned tensors, is that sufficient or do we need something else?

If quantized tensors should never be able to be pinned I can just disable that option. As is, this is more of a forward facing change, the only information that changes is device, the rest was just added because the functionality seemed like it might be useful in the future.

The code as is doesn't set the pinned_memory flag (see Tensor_Conversions.cpp function to_impl, pin_out is only used for non_quantized tensors) but it'd be theoretically possible to pin the memory if you called the empty_quantized function directly with that option set.

raghuramank100 reviewed (changes requested):
Please see comments in the PR, thanks

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).

HDCharles committed:
Update on "Cuda quantized tensors, support for quantize per tensor"


Summary: implements quantized tensors in cuda for for per_tensor
quantization, along with several necessary functions

Test Plan: python test/test_quantization.py TestQuantizedTensors
python test/test_quantization.py
TestQuantizedTensors.test_compare_quant_dequant_device_numerics
python test/test_quantization.py
TestQuantizedTensors.test_qtensor_to_device

Reviewers:

Subscribers:

Tasks:

Tags:

some fixes for the tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

fixing tests

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29018272](https://our.internmc.facebook.com/intern/diff/D29018272)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29018272).
